### PR TITLE
Fix duplicate defines when original value is not a dictionary

### PR DESCRIPTION
**Pull request by achilleas-k**
**otk: only try to merge dict with dict in defines**

If a key was previously defined in the context's defines but was not a
dict and was instead a list or primitive type, we would try to merge
them, which would break when trying to index the existing value as a
dictionary.

Make sure a subblock is reused only when it's actually a dictionary.  In
all other cases, the new value replaces the old.

---

**test: expand duplicate-define to include replacement cases**

The test now also checks if:
- a dictionary/object definition is replaced when the new value is a
  list or a string.
- a list or string definition is replaced when the new value is a
  dictionary/object.
- a list or string definition is replaced when the new value is a string
  or list (respectively).

---

### CODE REVIEW

**Review by supakeen — APPROVED** [review]
Fine to have them all in the same test; maybe rename it to `06-define-variants.yaml` or something and split out the duplicate define separately?

**Review by supakeen — APPROVED** [review]

**Review by mvo5 — COMMENTED** [review]
Thank you! The examples are excellent, I added a comment, we can merge as is and followup as this already helps and fixes the real world usage. but it would be nice to spend a bit more time here to see if we can clarify the connection of Context.vars and state a bit more.

File: src/otk/transform.py [hunk 1]
```diff
@@ -149,7 +149,12 @@ def process_defines(ctx: Context, state: State, tree: Any):
 
         if isinstance(value, dict):
             # the value is a dict: process it recursively under a new subblock
-            new_subblock = subblock.get(key, {})
+            new_subblock = {}
+            if isinstance(subblock.get(key), dict):
```
[the remark this answers is not in the document]
Comment (achilleas-k) [reply src/otk/transform.py:153]: I realise that the way I made the `subblock` a reference to the global context makes it hard to follow but it's kinda necessary for processing subkeys.
What we can do instead is put the key (nested if necessary) of the current subblock in State and use that to modify the context directly so it's more visible
so we would be doing something like:
```python
if isinstance(value, dict):
  ...
  new_state = state.copy(defines_key=key)
  process_includes(ctx, new_state, ...)
```
and then when we're adding a key, we do

```python
ctx.defines[state.defines_key] = new_value  # update the context directly
```

That would also allow us to go back to not exposing `defines` directly from the context and put it behind the methods that can also handle subkeys like `a.b`.

[the remark this answers is not in the document]
Comment (mvo5) [reply src/otk/transform.py:153]: Yeah, sounds fine for a followup given that this PR fixes a real issue in the centos examples. But I would love to see a bit more tweaks here after :)
